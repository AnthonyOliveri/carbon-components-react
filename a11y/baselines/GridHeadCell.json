{
    "issueMessages": {
        "messages": {
            "wcag20.tech.h57.htmlLangValid": "Html element must have lang and/or xml:lang attribute that conforms to BCP: 47.",
            "rpt.g481.htmlSkipNav": "Provide a mechanism to bypass blocks of content that are repeated on multiple Web pages.",
            "wcag20.tech.h25.docHasTitle": "There must be a non-empty title element in the head of the document.",
            "haac.g1132.BackgroundImgHasTextOrTitle": "Verify that important background image information is conveyed when the system is in high contrast mode.",
            "wcag20.tech.g1.bodyFirstASkips": "The page must provide a link to skip directly to the main page content as the first link on each page or use WAI-ARIA landmarks.",
            "rpt.g1.styleTrigger": "Web pages must be readable without requiring style sheets or equivalent facilitation is provided."
        },
        "lang": "en-us"
    },
    "reports": [
        {
            "issues": [
                {
                    "severityCode": "eISHigh",
                    "messageCode": "wcag20.tech.h25.docHasTitle",
                    "ruleId": "273",
                    "help": "idhi_accessibility_check_g273.html",
                    "msgArgs": [],
                    "xpath": "/html[1]",
                    "snippet": "<html style=\"\">",
                    "bounds": {
                        "left": 0,
                        "top": 0,
                        "height": 0,
                        "width": 0
                    },
                    "level": "violation"
                },
                {
                    "severityCode": "eISHigh",
                    "messageCode": "wcag20.tech.h57.htmlLangValid",
                    "ruleId": "490",
                    "help": "idhi_accessibility_check_g490.html",
                    "msgArgs": [],
                    "xpath": "/html[1]",
                    "snippet": "<html style=\"\">",
                    "bounds": {
                        "left": 0,
                        "top": 0,
                        "height": 0,
                        "width": 0
                    },
                    "level": "violation"
                },
                {
                    "severityCode": "eISHigh",
                    "messageCode": "wcag20.tech.g1.bodyFirstASkips",
                    "ruleId": "1046",
                    "help": "idhi_accessibility_check_g1046.html",
                    "msgArgs": [],
                    "xpath": "/html[1]/body[1]",
                    "snippet": "<body style=\"\">",
                    "bounds": {
                        "left": 0,
                        "top": 0,
                        "height": 0,
                        "width": 0
                    },
                    "level": "violation"
                }
            ],
            "frameIdx": 0
        }
    ],
    "summary": {
        "counts": {
            "violation": 3
        },
        "scanTime": 0,
        "policies": [
            "CI162_5_2_DCP070116",
            "CI162_5_2_DCP080115"
        ],
        "reportLevels": [
            "violation"
        ],
        "startScan": 1488578473743
    },
    "scanID": "5d45b399-69af-4e91-8ce6-b7d9ecf28d9a",
    "toolID": "@ibma/aat-v1.1.0",
    "label": "GridHeadCell"
}