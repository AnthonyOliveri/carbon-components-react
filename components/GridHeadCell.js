import React from 'react';
import PropTypes from 'prop-types';

export default class GridHeadCell extends React.Component {
  render() {
    const { column, sort } = this.props;
    const isSortable = column.sortable === true;
    const classes = ['bx--grid-head-cell', this.props.className];
    const titleClasses = ['bx--grid-head-cell__title'];
    let onSort = null;

    if (isSortable) {
      classes.push('bx--grid-head-cell--sortable');

      if (sort.property === column.name) {
        if (sort.direction === 'DESC') {
          titleClasses.push('bx--grid-head-cell__title--desc');
        } else {
          titleClasses.push('bx--grid-head-cell__title--asc');
        }
      }

      onSort = () => this.props.onSort(column);
    }

    const width = `${column.width}%`;

    return (
      <div
        className={classes.join(' ')}
        onClick={onSort}
        onDragOver={this.props.dragOver}
        style={{ width }}
      >
        <span className={titleClasses.join(' ')} key={column.title}>
          {column.title}
        </span>
        <div
          className="bx--grid-head-cell__handle"
          draggable="true"
          onDragEnd={e => this.props.dragEnd(e, column)}
          onDragStart={e => this.props.dragStart(e, column)}
        />
      </div>
    );
  }
}

GridHeadCell.defaultProps = {
  className: '',
  dragEnd: () => {},
  dragOver: () => {},
  dragStart: () => {},
  sort: {},
};

GridHeadCell.propTypes = {
  className: PropTypes.string,
  column: PropTypes.object.isRequired,
  dragEnd: PropTypes.func,
  dragOver: PropTypes.func,
  dragStart: PropTypes.func,
  onSort: PropTypes.func.isRequired,
  sort: PropTypes.shape({
    direction: PropTypes.string,
    property: PropTypes.string,
  }),
};
