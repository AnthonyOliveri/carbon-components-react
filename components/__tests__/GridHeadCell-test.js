import GridHeadCell from '../GridHeadCell';
import React from 'react';
import { shallow } from 'enzyme';

jest.unmock('../GridHeadCell');

const shallowRender = props => shallow(<GridHeadCell {...props} />);

describe('GridHeadCell', () => {
  describe('check if props are rendered correctly', () => {
    describe('standard cell - no sort', () => {
      const wrapper = shallowRender({
        className: 'class1',
        column: {
          name: 'column2',
          title: 'column2',
          render: data => data.column2,
          width: 25,
        },
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: jest.fn(),
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('standard cell - ASC sort', () => {
      const wrapper = shallowRender({
        className: 'class1',
        column: {
          name: 'column2',
          title: 'column2',
          sortable: true,
          render: data => data.column2,
          width: 25,
        },
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: jest.fn(),
        sort: {
          direction: 'ASC',
          property: 'column2',
        },
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('standard cell - DESC sort', () => {
      const wrapper = shallowRender({
        className: 'class1',
        column: {
          name: 'column2',
          title: 'column2',
          sortable: true,
          render: data => data.column2,
          width: 25,
        },
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: jest.fn(),
        sort: {
          direction: 'DESC',
          property: 'column2',
        },
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('check if events are triggered correctly', () => {
    describe('onSort is called correctly - if sortable is true', () => {
      const column = {
        name: 'column2',
        title: 'column2',
        sortable: true,
        render: data => data.column2,
        width: 25,
      };
      const onSortMock = jest.fn();
      const wrapper = shallowRender({
        className: 'class1',
        column,
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: onSortMock,
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('onSort is triggered on click', () => {
        wrapper.simulate('click');
        expect(onSortMock).toBeCalledWith(column);
      });
    });

    describe('onSort is not called - if sortable is false', () => {
      const column = {
        name: 'column2',
        title: 'column2',
        sortable: false,
        render: data => data.column2,
        width: 25,
      };
      const onSortMock = jest.fn();
      const wrapper = shallowRender({
        className: 'class1',
        column,
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: onSortMock,
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('onSort is triggered on click', () => {
        wrapper.simulate('click');
        expect(onSortMock).not.toBeCalled();
      });
    });

    describe('dragStart is called correctly', () => {
      const column = {
        name: 'column2',
        title: 'column2',
        render: data => data.column2,
        width: 25,
      };
      const onDragStartMock = jest.fn();
      const wrapper = shallowRender({
        className: 'class1',
        column,
        dragEnd: jest.fn(),
        dragOver: jest.fn(),
        dragStart: onDragStartMock,
        onSort: jest.fn(),
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('onDragStart is triggered', () => {
        wrapper.find('.bx--grid-head-cell__handle').simulate('dragStart', {
          event: 'x',
        });
        expect(onDragStartMock).toBeCalledWith(
          {
            event: 'x',
          },
          column
        );
      });
    });

    describe('dragEnd is called correctly', () => {
      const column = {
        name: 'column2',
        title: 'column2',
        render: data => data.column2,
        width: 25,
      };
      const onDragEndMock = jest.fn();
      const wrapper = shallowRender({
        className: 'class1',
        column,
        dragEnd: onDragEndMock,
        dragOver: jest.fn(),
        dragStart: jest.fn(),
        onSort: jest.fn(),
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('onDragEnd is triggered', () => {
        wrapper.find('.bx--grid-head-cell__handle').simulate('dragEnd', {
          event: 'x',
        });
        expect(onDragEndMock).toBeCalledWith(
          {
            event: 'x',
          },
          column
        );
      });
    });

    describe('dragOver is called correctly', () => {
      const column = {
        name: 'column2',
        title: 'column2',
        render: data => data.column2,
        width: 25,
      };
      const onDragOverMock = jest.fn();
      const wrapper = shallowRender({
        className: 'class1',
        column,
        dragEnd: jest.fn(),
        dragOver: onDragOverMock,
        dragStart: jest.fn(),
        onSort: jest.fn(),
        sort: {
          direction: 'ASC',
          property: 'column3',
        },
      });

      it('onDragOver is triggered', () => {
        wrapper.simulate('dragOver');
        expect(onDragOverMock).toBeCalled();
      });
    });
  });
});
