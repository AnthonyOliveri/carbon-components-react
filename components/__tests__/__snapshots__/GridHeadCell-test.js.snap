exports[`GridHeadCell check if props are rendered correctly standard cell - ASC sort renders correctly 1`] = `
<div
  className="bx--grid-head-cell class1 bx--grid-head-cell--sortable"
  onClick={[Function]}
  onDragOver={[Function]}
  style={
    Object {
      "width": "25%",
    }
  }>
  <span
    className="bx--grid-head-cell__title bx--grid-head-cell__title--asc">
    column2
  </span>
  <div
    className="bx--grid-head-cell__handle"
    draggable="true"
    onDragEnd={[Function]}
    onDragStart={[Function]} />
</div>
`;

exports[`GridHeadCell check if props are rendered correctly standard cell - DESC sort renders correctly 1`] = `
<div
  className="bx--grid-head-cell class1 bx--grid-head-cell--sortable"
  onClick={[Function]}
  onDragOver={[Function]}
  style={
    Object {
      "width": "25%",
    }
  }>
  <span
    className="bx--grid-head-cell__title bx--grid-head-cell__title--desc">
    column2
  </span>
  <div
    className="bx--grid-head-cell__handle"
    draggable="true"
    onDragEnd={[Function]}
    onDragStart={[Function]} />
</div>
`;

exports[`GridHeadCell check if props are rendered correctly standard cell - no sort renders correctly 1`] = `
<div
  className="bx--grid-head-cell class1"
  onClick={null}
  onDragOver={[Function]}
  style={
    Object {
      "width": "25%",
    }
  }>
  <span
    className="bx--grid-head-cell__title">
    column2
  </span>
  <div
    className="bx--grid-head-cell__handle"
    draggable="true"
    onDragEnd={[Function]}
    onDragStart={[Function]} />
</div>
`;
